File: bytopic/math/java/group3/MaxPointsOnALine.java
// Given an array of points where points[i] = [xi, yi] represents a point on the X-Y plane, return the maximum number
// of points that lie on the same straight line.
// Example:
// Input: points = [[1,1],[3,2],[5,3],[4,1],[2,3],[1,4]]
// Output: 4
// Note: Visualization may be needed to understand
//
// Tag: 136/150
// Tag: 149/2927, R151/2936 (overall frequency ranking)
// Note: Program compiles but output is not correct, debug and find out.


import java.util.HashMap;
import java.util.Map;
class MaxPointsOnALine {
    public static void main(String[] args) {
        int[][] points = {{1,1},{3,2},{5,3},{4,1},{2,3},{1,4}};
        System.out.println("Max. points on a line: " + maxPoints(points));
    }

    static int maxPoints(int[][] points) {
        int n = points.length;
        if (n <= 2) return n;
        int ans = 2;

        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int temp = 2;
                for (int k = j + 1; k < n; k++) {
                    int x = (points[j][1] - points[i][1] * (points[k][0] - points[i][0]));
                    int y = (points[k][1] - points[i][1]) * (points[j][0] - points[i][0]);
                    if (x == y) {
                        temp++;
                    }
                }
                if (temp > ans) {
                    ans = temp;
                }
            }
        }

        return ans;
    }
}

File: bytopic/math/python/group3/max_points_on_a_line.py
"""
Given an array of points where points[i] = [xi, yi] represents a point on the X-Y plane, return the maximum number
of points that lie on the same straight line.

Example:
Input: points = [[1,1],[3,2],[5,3],[4,1],[2,3],[1,4]]
Output: 4
Note: Visualization may be needed to understand

Tag: 136/150
Tag: 149/2927, R289/2936 (overall frequency ranking)
"""
from collections import defaultdict
from math import gcd


def max_points(points: list[list[int]]) -> int:
                                            #   points = [[1,0],[2,1],[3,4], [5,4]]

    points.sort()                           #   point1 point2 (dx,dy)    m     slope               mm
    slope, mm = defaultdict(int), 0         #   –––––– –––––– ––––––– –––––––  –––––––––––        –––––
                                            #   [1,0]  [2,1]   (1,1)   (1,1)   {(1,1):1}            1
    for i, (x1, y1) in enumerate(points):   #          [3,4]   (2,4)   (1,2)   {(1,1):1,(1,2):1}    1
                                            #          [5,4]   (4,4)   (1,1)   {(1,1):2,(1,2):1}    2
        slope.clear()                       #   [2,1]  [3,4]   (1,3)   (1,3)   {(1,3):1}            2
                                            #          [5,4]   (3,3)   (1,1)   {(1,3):1,(1,1):1}    2
        for x2, y2 in points[i + 1:]:       #   [3,4]  [5,4]   (3,0)   (1,0)   {(1,0):1}            2
            dx, dy = x2 - x1, y2 - y1
                                            #  mm + 1 = 2 + 1 = 3 <-- return
            gg = gcd(dx, dy)
            m = (dx//gg, dy//gg)

            slope[m] += 1
            if slope[m] > mm:
                mm = slope[m]

    return mm + 1


def main():
    points = [[1, 1], [3, 2], [5, 3], [4, 1], [2, 3], [1, 4]]
    print(max_points(points))


if __name__ == "__main__":
    main()
